### PR TITLE
api/client: improve constructor configurability

### PR DESCRIPTION
**Pull request by tychoish**
I think being able to configure these options will make it easier to
reuse existing HTTP request infrastructure in our application and call
fly.io APIs from within our application.

### CODE REVIEW

File: api/client.go [hunk 1]
```diff
@@ -42,23 +42,60 @@ type Client struct {
...
+}
 
-	url := fmt.Sprintf("%s/graphql", baseURL)
+func NewClientFromOptions(opts ClientOptions) *Client {
+	var transport *Transport
+	if opts.Transport == nil {
+		transport = opts.Transport
+	} else {
+		transport = &Transport{}
+	}
```
[the remark this answers is not in the document]
Comment (tychoish) [reply api/client.go]: Ah good catch. Sorry, I got the if statement backwards. I had/have a slight bias towards not constructing the `Transport` object twice in the case that we run the statement on line 90. 

There aren't any existing tests in this package, but I'm happy to add one.

**Comment by tychoish** [comment]
> Can you also shed more light on how changing how we setup our transport helps your application?

I was mostly hoping to be able to use an `http.Client` instance that was connected to our instance. I also saw the context in the transport and wanted to pass our own in; however, I now see that this is unused and likely vestigial.

File: api/client.go [hunk 2]
```diff
@@ -42,23 +42,58 @@ type Client struct {
...
+		t.EnableDebugTrace = !slices.Contains([]string{"", "0", "false"}, os.Getenv("FLY_FORCE_TRACE"))
+	}
+}
 
-	url := fmt.Sprintf("%s/graphql", baseURL)
+func NewClientFromOptions(opts ClientOptions) *Client {
+	transport := &Transport{}
+	if opts.Transport != nil {
+		transport = opts.Transport
+	}
```
Comment (dangra) [inline api/client.go:88]: a small twist suggestion:

```go
transport := opts.Transport
if transport == nil {
  transport = &Transport{}
}
```

**Comment by dangra** [comment]
needs to resolve the conflicts to merge it

File: api/client.go (hunk 2 above)
Comment (tychoish) [reply api/client.go]: ah good call!

**Comment by dangra** [comment]
thank you!

**Pull request merged**